realtek: pcs: some SerDes mode handling improvements

jonasjelonek opened the pull request:
This includes various cleanups, improvements and fixes for the RTL931X SerDes setup, most notably making use of previously introduced groundwork to improve the quality and clarity of the code.

Note: the diff of commit c896e4670329a3e51df7b1424920f69c4cb02fcc sadly looks quite weird and may be hard to review, though it's just a code move.

jonasjelonek committed:
realtek: pcs: rtl931x: don't return error on USXGMII setup

Configuring USXGMII on RTL931X was disable because it is not setup
properly right now. Affected devices need to rely on U-boot setup for
now. However, it is not the proper way to return an error in case we
just want to skip USXGMII configuration. Thus, just return 0 in that
case indicating it is properly configured.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: move SerDes config out of setup

Slim the rtpcs_931x_setup_serdes entrypoint by moving the SerDes
configuration sequences for the different SerDes modes into a dedicated
function called rtpcs_931x_sds_config_mode. They form a logical block
similar to what the SDK does and can be taken out to follow the
'divide & conquer' principle.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: separate and reuse 1000base-x config

Move the configuration of 1000Base-X mode into its own function. This
way, redundant code can be replaced by a single function call.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: map interface mode to hardware mode

The PCS driver is intended to use a dedicated hardware mode in the
future because the hardware capabilities and modes of the SerDes do not
align with the PHY_INTERFACE_MODE_* modes.

For the beginning, use the previously introduced generic mode mapping
in the SerDes setup and assign the determined hardware mode to the
corresponding field of a SerDes instance. This allows parallel usage of
both mode representations for the time being, until the phy_interface_t
modes can be dropped from internal functions completely.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: use hardware mode for configuration

Pass the SerDes hardware mode to rtpcs_931x_sds_config_mode instead of
the PHY_INTERFACE_MODE_* mode and use it. Replace matched modes
accordingly.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: reorder modes in mode configuration

Reorder the modes within the switch statement in
rtpcs_931x_sds_config_mode to have a proper and non-confusing order.
Have fiber modes first and MII modes after, and sort all modes in
ascending speed/capabilities order.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: rtl931x: add helper for SerDes power up/down

Add a dedicated helper 'rtpcs_931x_sds_power' to power down/up a SerDes.
Up to now, plain register writes to the corresponding register were used
in several places and with slightly different syntax. Use the helper to
replace the existing boilerplate code for powering on/off SerDes.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek commented on target/linux/realtek/files-6.12/drivers/net/pcs/pcs-rtl-otto.c:
```diff
@@ -2817,6 +2818,33 @@ static int rtpcs_931x_setup_serdes(struct rtpcs_serdes *sds,
+	default:
+		dev_err(ctrl->dev, "SerDes %u doesn't support %s mode\n", sds_id, phy_modes(mode));
+		return -ENOTSUPP;
+	}
```
It's one of my current goals, besides making the code better overall. But I guess our definitions of "better code" and the way to achieve this slightly differ.

Though, I'm fine with this and will provide this as a generic mapper.

plappermaul replied:
Every effort to make the realtek world better is a good effort.

jonasjelonek committed:
realtek: pcs: rename SerDes mode field

Rename the mode field in rtpcs_serdes from 'mode' to 'hw_mode' to make
clear what it actually is, to avoid confusion with the phy_interface_t
'mode' usually passed to functions.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

jonasjelonek committed:
realtek: pcs: add mapping for SerDes hardware mode

Add a generic mapper for mapping the PHY_INTERFACE_MODE_* modes to the
SerDes hardware modes RTPCS_SDS_MODE_*. This is to be used by all
subtargets step-by-step and avoids having subtarget-specific code for
that simple mapping.

Signed-off-by: Jonas Jelonek <jelonek.jonas@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/21184
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

plappermaul reviewed (changes requested):
Only two comments left here ... otherwise ready to go.
